Introduce `PINFuturePending` for handling delegate patterns

raycsh017 opened the pull request:
This commit introduces `PINFuturePending`, a wrapper on top of `PINFuture`
that could be used to handle cases where we expect the future to be resolved
later in some other part of the code, e.g. 3rd party delegate method that
we have little control over when it will get called and resolve or reject the
future based on the result. The benefit of this is that we'll be able to chain
success/failure methods right away to the future without having to wait for it
to resolve or get rejected.

bolsinga commented on Example/Pods/Pods.xcodeproj/project.pbxproj:
```diff
@@ -43,7 +43,10 @@
+		4C666DC22400783E006F21B9 /* PINFuturePending.m in Sources */ = {isa = PBXBuildFile; fileRef = 4CD0835623FF5C65003273DE /* PINFuturePending.m */; };
+		4C666DC32400784E006F21B9 /* PINFuturePending.h in Headers */ = {isa = PBXBuildFile; fileRef = 4CD0835723FF5C66003273DE /* PINFuturePending.h */; };
 		4CA4F8F64565882DCFA174872749C0EF /* EXPFloatTuple.h in Headers */ = {isa = PBXBuildFile; fileRef = 17B271C769948413B2D6A2FA5FA661F5 /* EXPFloatTuple.h */; settings = {ATTRIBUTES = (Public, ); }; };
+		4CD0835923FF5C66003273DE /* PINFuturePending.h in Headers */ = {isa = PBXBuildFile; fileRef = 4CD0835723FF5C66003273DE /* PINFuturePending.h */; };
```
I have a hunch this needs to be made into a Public header file.

jparise commented on PINFuture/Classes/PINFuturePending.m:
```diff
@@ -0,0 +1,49 @@
+    if (self) {
+        _pendingFuture = [PINFuture withBlock:^(void (^_Nonnull resolve)(id), void (^_Nonnull reject)(NSError *_Nonnull)) {
+            self.resolve = [resolve copy];
+            self.reject = [reject copy];
```
Is the `-copy` necessary if the properties are themselves marked as `copy`?

raycsh017 committed:
Introduce `PINFuturePending` for handling delegate patterns

This commit introduces `PINFuturePending`, a wrapper on top of `PINFuture`
that could be used to handle cases where we expect the future to be resolved
later in some other part of the code, e.g. 3rd party delegate method that
we have little control over when it will get called and resolve or reject the
future based on the result. The benefit of this is that we'll be able to chain
success/failure methods right away to the future without having to wait for it
to resolve or get rejected.

jparise commented:
@raycsh017 do you think this should be mentioned in the documentation (`README.md`)?

jparise commented:
@raycsh017 This could also use a unit test before we merge it.

garrettmoon commented:
> @raycsh017 do you think this should be mentioned in the documentation (`README.md`)?

I second this, with a specific example of what this is good for.

raycsh017 committed:
Add unit test for `PINFuturePending` and reinforce methods for it

This commit adds unit tests for testing basic functionality of
`PINFuturePending`, as well as testing for common error cases users
might run into (e.g. resolving twice).

Also, this commit reinforces the methods for the new class to better
handle error cases and avoid reference cycles.

Finally, for some reason the file was missing from the project file,
this commit adds it back.

raycsh017 committed:
Update project file with changes that come with new build changes(?)

This commit is for saving the changes made by Xcode building
`PINFuture-Example` scheme.

jparise commented on PINFuture/Classes/PINFuturePending.m:
```diff
@@ -0,0 +1,65 @@
+    if (self.resolve) {
+        self.resolve(value);
+    } else {
+        NSAssert(NO, @"Underlying future has already been resolved.");
```
It might be nicer to write this above the `if` (rather than in an `else` block) as:

```objc
NSAssert(self.resolve, @"Underlying future has already been resolved.");
```

... although the way you have it makes it easier to set a breakpoint on the error condition.

raycsh017 replied:
That sounds good. I just personally like having `else` cases to be clearer that we are handling all cases

jparise reviewed (commented):
It would still be nice to mention this in the `README.md`, too.

raycsh017 committed:
Update README.md to include `PINFuturePending`

raycsh017 committed:
Move asserts above if statements in `PINFuturePending`

raycsh017 commented:
@jparise @garrettmoon Updated `README.md` to include when and how to use `PINFuturePending`. Let me know if any part could be clearer :)

jparise reviewed (approved):
Looks good to me, but let's get one of the other folks to give this a 👍, too.

raycsh017 commented:
@garrettmoon @rahul-malik Could you guys take a look when you get the chance? Added documentation around the new class as well as unit tests!

rahul-malik commented on PINFuture/Classes/PINFuturePending.m:
```diff
@@ -0,0 +1,63 @@
+{
+    NSParameterAssert(error != nil);
+    NSParameterAssert([error isKindOfClass:[NSError class]]);
+    NSAssert(self.reject, @"Underlying future has already been resolved.");
```
shouldn't the message be about rejection?

raycsh017 replied:
Nope, since this could catch the case where you fulfill the pending future and try to call reject on the same future again.

bolsinga commented on README.md:
```diff
@@ -252,6 +252,53 @@ userFuture = [userFuture executor:[PINExecutor main] chainSuccess:^(User *user)
+
+    // Start sign in flow
+    GIDSignIn *signIn = [GIDSignIn sharedInstance];
+    signIn.delegate = self;
```
What if the PINFuture here was instead a PINFuture subclass, and it conformed to the GIDSIgnIn delegate itself? I'll contact you offline.

rahul-malik reviewed (approved):
@raycsh017 - Seems reasonable to me, the main concern I have is if the right solution here is another type or if we just need to create a new way without blocks for resolving / rejecting futures

I'll unblock this for now until I have a better solution in mind
